Repeated --require/--import in NODE_OPTIONS is corrupted when forking workers

### Link to the code that reproduces this issue

https://github.com/colinhacks/nextjs-node-options-repro

### To Reproduce

```bash
git clone https://github.com/colinhacks/nextjs-node-options-repro
cd nextjs-node-options-repro
npm install
npm run repro
```

The script runs `next build` three times, varying only `NODE_OPTIONS`, and reports how many processes each preload actually ran in.

### Current vs. Expected behavior

Node accepts a repeated `--require` and runs every one of them. Next reparses `NODE_OPTIONS` and reformats it for each forked worker, and that round-trip cannot represent two of the same flag.

Exact process counts vary slightly per run; the pattern does not.

| Case | `NODE_OPTIONS` | Outcome |
|---|---|---|
| 1 (control) | `--require=<a>` | Preload a runs in every process. Exit 0. |
| 2 | `--require=<a> --require=<b>` | Preload a is missing from the forked workers — b runs in all of them, a in only a few. **Exit 0.** |
| 3 | `--require <a> --require <b>` | Build fails, exit 1. |

Case 3 fails with both paths joined into one specifier:

```
Error: Cannot find module '/path/preload-a.cjs /path/preload-b.cjs'
  code: 'MODULE_NOT_FOUND',
  requireStack: [ 'internal/preload' ]
```

Case 2 is the more dangerous one, since the build succeeds and the missing preload is silent.

Expected in both: each preload runs in every process, as it does under plain `node`.

### Root cause

Parsing in `server/lib/utils.ts` stores Node options in a `Record` keyed by option name:

```ts
export type NodeOptions = Record<string, string | boolean | undefined>

const parseNodeArgs = (args: string[]): NodeOptions => {
  const { values, tokens } = parseArgs({ args, strict: false, tokens: true })
```

A `Record` cannot hold two `require` entries, and the two spellings lose the data differently.

The equals form keeps only the last value, because `parseArgs` with `strict: false` and no `options` config does not accumulate:

```js
parseArgs({ args: ["--require=A", "--require=B"], strict: false }).values
// { require: 'B' }
```

The space-separated form instead reaches the orphan branch, which concatenates:

```ts
values[orphan.name] += ` ${token.value}`
```

producing the single bogus specifier `A B`. `formatNodeOptions` then rebuilds the string from that `Record`, and `getFormattedNodeOptionsWithoutInspect` is what each forked worker receives.

### Suggested fix

Declaring the repeatable flags as arrays preserves them:

```js
parseArgs({
  args: ["--require=A", "--require=B"],
  strict: false,
  options: { require: { type: "string", multiple: true } },
}).values
// { require: [ 'A', 'B' ] }
```

`formatNodeOptions` would then emit one flag per array element. The same applies to `--import`, which reproduces identically.

### Provide environment information

```
Operating System:
  Platform: darwin
  Arch: arm64
  Available memory (MB): 65536
  Available CPU cores: 10
Binaries:
  Node: 26.5.0
  npm: 11.17.0
  pnpm: 10.15.1
Relevant Packages:
  next: 16.3.0 // Latest available version is detected (16.3.0).
  react: 19.2.0
  react-dom: 19.2.0
  typescript: 7.0.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Not sure

### Which stage(s) are affected? (Select all that apply)

next build (local), next dev (local)

### Additional context

Affects any setup needing more than one preload — an APM agent alongside an env loader, or a runtime that injects its own `--require` and lets the user add their own.

Related, and the reason this is filed fresh rather than as a comment:

- #77550 is open and covers case 3 (the space-form crash) via New Relic plus dotenv. It does not cover case 2, the silent drop, and no root cause was identified there. I have also left this analysis on that thread.
- #67286 reported the same bug in 2024. It was closed by the stale bot with no human response and then locked; that bot message invites a new issue.

--- a/packages/next/src/server/lib/utils.test.ts
+++ b/packages/next/src/server/lib/utils.test.ts
@@ -77,6 +77,19 @@ describe('formatNodeOptions', () => {
       ],
     })
   })
+
+  it('emits one flag per element for array values', () => {
+    const result = formatNodeOptions({
+      require: ['/path/a.cjs', '/path/b.cjs'],
+      import: '/path/c.mjs',
+    })
+
+    expect(result).toEqual({
+      execArgv: [],
+      nodeOptions:
+        '--require=/path/a.cjs --require=/path/b.cjs --import=/path/c.mjs',
+    })
+  })
 })
 
 describe('getMemoryRestartStats', () => {
@@ -191,5 +204,38 @@ describe('getFormattedNodeOptionsWithoutInspect', () => {
     const result = getFormattedNodeOptionsWithoutInspect()
 
     expect(result).toBe('--other --inspect-port=0.0.0.0:1234 --additional')
+  })
+
+  it('preserves repeated --require options', () => {
+    process.env.NODE_OPTIONS =
+      '--require=/path/a.cjs --require=/path/b.cjs --other'
+    const result = getFormattedNodeOptionsWithoutInspect()
+
+    expect(result).toBe('--require=/path/a.cjs --require=/path/b.cjs --other')
+  })
+
+  it('preserves repeated --require options with space-separated values', () => {
+    process.env.NODE_OPTIONS = '--require /path/a.cjs --require /path/b.cjs'
+    const result = getFormattedNodeOptionsWithoutInspect()
+
+    expect(result).toBe('--require=/path/a.cjs --require=/path/b.cjs')
+  })
+
+  it('preserves repeated --import options', () => {
+    process.env.NODE_OPTIONS =
+      '--import=/path/a.mjs --import=/path/b.mjs --other'
+    const result = getFormattedNodeOptionsWithoutInspect()
+
+    expect(result).toBe('--import=/path/a.mjs --import=/path/b.mjs --other')
+  })
+
+  it('preserves repeated --require with quoted values', () => {
+    process.env.NODE_OPTIONS =
+      '--require "./a with spaces.js" --require "./b with spaces.js"'
+    const result = getFormattedNodeOptionsWithoutInspect()
+
+    expect(result).toBe(
+      '--require="./a with spaces.js" --require="./b with spaces.js"'
+    )
   })
 })

--- a/packages/next/src/server/lib/utils.ts
+++ b/packages/next/src/server/lib/utils.ts
@@ -11,10 +11,32 @@ export function printAndExit(message: string, code = 1) {
   return process.exit(code)
 }
 
-export type NodeOptions = Record<string, string | boolean | undefined>
+export type NodeOptions = Record<
+  string,
+  string | string[] | boolean | undefined
+>
+
+/**
+ * Node.js CLI flags that can be repeated. Node runs every occurrence of these
+ * flags, so they need to be preserved individually rather than collapsed into
+ * a single value keyed by option name.
+ * https://nodejs.org/api/cli.html#--requiremodule
+ * https://nodejs.org/api/cli.html#--importmodule
+ */
+const REPEATABLE_OPTIONS: Record<string, { type: 'string'; multiple: true }> = {
+  require: { type: 'string', multiple: true },
+  import: { type: 'string', multiple: true },
+}
 
 const parseNodeArgs = (args: string[]): NodeOptions => {
-  const { values, tokens } = parseArgs({ args, strict: false, tokens: true })
+  const parsed = parseArgs({
+    args,
+    strict: false,
+    tokens: true,
+    options: REPEATABLE_OPTIONS,
+  })
+  const values: NodeOptions = parsed.values
+  const tokens = parsed.tokens
 
   // For the `NODE_OPTIONS`, we support arguments with values without the `=`
   // sign. We need to parse them manually.
@@ -46,12 +68,27 @@ const parseNodeArgs = (args: string[]): NodeOptions => {
       continue
     }
 
+    const name = orphan.name
+
+    // If the option is repeatable, accumulate each occurrence instead of
+    // collapsing them into a single value.
+    if (name in REPEATABLE_OPTIONS) {
+      if (Array.isArray(values[name])) {
+        values[name].push(token.value)
+      } else if (typeof values[name] === 'string') {
+        values[name] = [values[name], token.value]
+      } else {
+        values[name] = token.value
+      }
+      continue
+    }
+
     // If the token is a positional one, and it has a value, so add it to the
     // values object. If it already exists, append it with a space.
-    if (orphan.name in values && typeof values[orphan.name] === 'string') {
-      values[orphan.name] += ` ${token.value}`
+    if (name in values && typeof values[name] === 'string') {
+      values[name] += ` ${token.value}`
     } else {
-      values[orphan.name] = token.value
+      values[name] = token.value
     }
   }
 
@@ -150,7 +187,7 @@ export const formatDebugAddress = ({ host, port }: DebugAddress): string => {
  * @returns An object with the host and port of the debug address.
  */
 export const getParsedDebugAddress = (
-  address: string | boolean | undefined
+  address: string | string[] | boolean | undefined
 ): DebugAddress => {
   if (!address || typeof address !== 'string') {
     return { host: undefined, port: 9229 }
@@ -183,10 +220,21 @@ const EXEC_ARGV_ONLY_OPTIONS = new Set([
 
 function formatArg(
   key: string,
-  value: string | boolean | undefined
+  value: string | string[] | boolean | undefined
 ): string | null {
   if (value === true) {
     return `--${key}`
+  }
+
+  if (Array.isArray(value)) {
+    const formatted: string[] = []
+    for (const item of value) {
+      const formattedItem = formatArg(key, item)
+      if (formattedItem !== null) {
+        formatted.push(formattedItem)
+      }
+    }
+    return formatted.length > 0 ? formatted.join(' ') : null
   }
 
   if (value) {
@@ -210,9 +258,10 @@ function formatArg(
  * @param args The arguments to be stringified.
  * @returns An object with `nodeOptions` string and `execArgv` array.
  */
-export function formatNodeOptions(
-  args: Record<string, string | boolean | undefined>
-): { nodeOptions: string; execArgv: string[] } {
+export function formatNodeOptions(args: NodeOptions): {
+  nodeOptions: string
+  execArgv: string[]
+} {
   const nodeOptionsParts: string[] = []
   const execArgv: string[] = []
 
@@ -230,10 +279,7 @@ export function formatNodeOptions(
   return { nodeOptions: nodeOptionsParts.join(' '), execArgv }
 }
 
-export function getParsedNodeOptions(): Record<
-  string,
-  string | boolean | undefined
-> {
+export function getParsedNodeOptions(): NodeOptions {
   const args = [...process.execArgv, ...getNodeOptionsArgs()]
   if (args.length === 0) return {}
 

